NDTensors/src/diag/tensoralgebra/contract.jl

```diff
@@ -129,7 +129,7 @@ function contract!(
       # Assumes C starts set to 0
       c₁ = zero(ElC)
       for i in 1:min_dim
-        c₁ += getdiagindex(A, i) * getdiagindex(B, i)
+        c₁ += getdiagindex(expose(A), i) * getdiagindex(expose(B), i)
```

Comment:
Instead of this approach I think it would be better to catch cases when dense * diag contractions are occurring on GPU (say with `expose`) and the convert the diag tensor to dense and then perform a dense contraction.

Reply:
I will get started on that! I was actually surprised that this `@allowscalar` was still faster than the CPU version. Thanks!

Reply:
Yeah that is surprising... It was good that you tried it out but I think converting to dense will be more robust.

Reply:
@mtfishman I wrote a draft that fixes the problem and goes through `contract!`. Right now, because `delta` is by default constructed on CPU, converting the `DiagTensor` to `DenseTensor` constructs it as a CPU array. Right now I am using adapt to move it to GPU but this will be updated with the changes in `delta`. I am working on adding a test now

Reply:
You say:
> Right now, because delta is by default constructed on CPU, converting the DiagTensor to DenseTensor constructs it as a CPU array.
I don't really understand what you mean. Which delta is being constructed on CPU? Do you mean the ones in the `tr` operation?

Reply:
Sorry that was confusing I updated the comment. I am talking about the delta in trace the operation 
```
Tᶜ *= δ(eltype(T), dag((cᴸ, cᴿ)))
```
Converting this delta or any `UniformDiag` to `DenseTensor` using `dense` uses `Vector` as the datatype of the new `Dense` object. See this example
```
d = tensor(δ(eltype(T), dag((cᴸ, cᴿ))))
typeof(d) = NDTensors.DiagTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Diag{Float32, Float32}}

typeof(dense(d)) = NDTensors.DenseTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Vector{Float32}}}
```

Reply:
Makes sense, thanks for the clarification.

Reply:
So what is your proposed plan for constructing those tensors on GPU when that is needed? I can imagine that happening in different parts of the code, with potential costs and benefits to those different choices.

Reply:
Yes I totally agree and I do not yet have a solid plan. I am going to try out different solutions and will let you know when I have more ideas!